### ISSUE TITLE
Only flag CSP that is actually less strict than the default

### ISSUE THREAD

**Issue by wagnerand** [mozilla/addons-linter#1298]
We currently flag any definition of CSP in the manifest.
This causes quite some false positives, and for post-review, we need it to be more reliable.

We are mostly interested in remote scripts, other remote resources should be fine.
So, I *think* what we need to check for is whether `http` or `*` is in `script-src` or `default-src`.

@jonathanKingston @marumari could you confirm what I wrote above is correct and complete? Thanks!

**Comment by april** [comment]
It's actually more complicated than that, as you are probably unsurprised by.

Are you not wanting to allow *any* remote scripts, or want to just allow scripts over https?

**Comment by wagnerand** [comment]
@april Sorry for using your wrong github profile.

Our policies do not allow remote scripts at all, so we'd want to flag *any* remote script.

**Comment by april** [comment]
It's okay, I actually just switched from @marumari to @april; I just re-reserved marumari so that nobody uses it to phish as me.

If you want a bulletproof solution, then having a CSP with `script-src` set to `moz-extension://{{UUID}}/` is probably going to be the only way to successfully enforce it.

If you're just trying to catch a CSP for somebody who is unaware of policies, I *think* blocking something like looking for `*`, `http:`, `https:`, and `ftp:` should work. There may be other schemes that one can load script over, I'd check with #contentsecurity to be sure.

If you're trying to stop somebody from defeating your scanning systems, that will be harder.  You can look for things like `createElement('script')` to catch script fetched over XHR, but you have to watch handlers, JSONP, etc. etc.

**Comment by wagnerand** [comment]
I'd like to get this in for MVP as well.

`<@cgrebs> parsing them is straight forward, agreed`

**Comment by erosman** [comment]
Note: Not all remote scripts are actual LINKs. They can be a sha256 hash)
`"content_security_policy": "script-src 'self' 'sha256-qznLcsROx4GACP2dm0UCKCzCG+HiZ1guq6ZZDob/Tng='; object-src 'self'"`

Here is just a quick idea .... (in JavaScript)

```
// parse manifest.json
let manifest = JSON.parse(str);

// get "content_security_policy"
let CSP = manifest["content_security_policy"];
if (CSP) { 
  
  // we are happy with 'self'
  CSP = CSP.replace(/'self'/g, '');
  
  // check script-src
  let scriptSrc = CSP.match(/script\-src\s+([^\s;]+)/);
  if (scriptSrc) { 
    console.log(scriptSrc[1]);
    // it has something that we need to look at
  }
  
  let defaultSrc = CSP.match(/default\-src\s+([^\s;]+)/);
  if (defaultSrc) { 
    console.log(defaultSrc[1]);
    // it has something that we need to look at
  }
} 
```

**Comment by erosman** [comment]
**Note on sha256 hash**

AFA I understand, using a sha256 hash will allow any source with that hash (local or remote). In fact it makes little sense in using a hash for a local script.

**Comment by april** [comment]
Hashes are for only inline script, they don't work for remote script.  :)

**Comment by erosman** [comment]
Is that to use as an alternative for 'unsafe-inline'?

**Comment by april** [comment]
Yes, it's essentially "safe" inline.  :)

**Comment by erosman** [comment]
Thank you .. I take it back :)

**Comment by wagnerand** [comment]
Thank you @april for clarifying!

**Comment by erosman** [comment]
Just for reference:
https://www.w3.org/TR/2015/CR-CSP2-20150721/#script-src-hash-usage

**Comment by ValentinaPC** [comment]
Please add some STR for us to check this or the label "qa: not needed". Thanks!

**Comment by jonathanKingston** [comment]
@april do you know if we block nonces from being used in addons? If not we certainly should reject them at review time.

**Comment by april** [comment]
I'm not 100% certain about the add-on architecture, but if a background script wanted to inject a script into the page's context, it could inject a nonce into the CSP policy, then inject the script with that nonce. That's pretty edge-case, and only would apply to dynamic nonces.

It certainly wouldn't apply to a manifest since that's not really a nonce, more of a number used forever.
